mp/src/game/shared/momentum/mom_gamemovement.cpp

```diff
@@ -1286,20 +1402,43 @@ void CMomentumGameMovement::CategorizePosition()
                     }
                 }
 
-                ClipVelocity(vecNextVelocity, pm.plane.normal, vecNextVelocity, 1.0f);
-
-                // Set ground entity if the player is not going to slide on a ramp next tick and if they will be
-                // grounded (exception if the player wants to bhop)
-                if (vecNextVelocity.z <= NON_JUMP_VELOCITY && bGrounded)
+                if (bGrounded)
                 {
-                    // Make sure we check clip velocity on slopes/surfs before setting the ground entity and nulling out
-                    // velocity.z
-                    if (sv_slope_fix.GetBool() && vecNextVelocity.Length2DSqr() > mv->m_vecVelocity.Length2DSqr())
+                    if (sv_slope_fix.GetBool())
                     {
-                        VectorCopy(vecNextVelocity, mv->m_vecVelocity);
-                    }
+                        ClipVelocity(vecNextVelocity, pm.plane.normal, vecNextVelocity, 1.0f);
+
+                        // What constitutes a favorable landing depends on if the player is allowed to bhop.
+
+                        // It also depends on if ground speed is capped *and* if the player actually bhops
+                        // on the next tick, but if we assume ground speed is capped only when bhopping isn't
+                        // allowed, all that matters is if bhopping is allowed.
+
+                        // On modes that can't bhop, colliding before landing is better if it means they start sliding.
+                        if (g_pGameModeSystem->GetGameMode()->CanBhop() || vecNextVelocity.z <= NON_JUMP_VELOCITY)
+                        {
+                            // Only update velocity as if we collided if it results in horizontal speed gain.
+                            // Otherwise, we are probably going uphill and are actually trying to avoid this collision.
+                            if (vecNextVelocity.Length2DSqr() > mv->m_vecVelocity.Length2DSqr())
+                            {
+                                VectorCopy(vecNextVelocity, mv->m_vecVelocity);
+                            }
 
-                    SetGroundEntity(&pm);
+                            SetGroundEntity(&pm);
+                        }
+                        else
+                        {
+                            // Pretend the player collided first and since that means they can't land,
+                            // don't set the ground entity
+
+                            // Should we also pre-apply the ClipVelocity result?
+                            // VectorCopy(vecNextVelocity, mv->m_vecVelocity);
+                        }
```

Comment:
Empty block? You can remove this and leave the comment to explain why there's no else.

Reply:
I'll do that now. Wasn't totally set on if that copy operation should actually be there or not but it should be fine leaving it out.